Match retail BSS ordering

cadmic opened the pull request:
Strategies used:
* Declaring some variables in header vs only in the .c file. In -g3 it seems like the block number is assigned the first time a variable is declared. For these I renamed them from `s` to `g` (although none of them were technically static before and I'm not sure if that matters).
* Adding dummy structs, using the new macros `INCREMENT_BLOCK_NUMBER_BY_1()`, `INCREMENT_BLOCK_NUMBER_BY_10()`, and `INCREMENT_BLOCK_NUMBER_BY_100()`
* Moving variables and struct definitions around 
* Adding or removing scopes inside functions by moving local variables around

For some files I left a comment to hopefully make it easier to fix them if headers change. It's hard to do in general though, especially for files which have variables declared both in a header and in the .c file. It wasn't too bad to fix though with the new tooling and [a spreadsheet](https://docs.google.com/spreadsheets/d/17OsxtmSJVX9dvzW-CQFVaeL4YSYWc2R6SZxcg1I7pac/edit#gid=1452764857) so hopefully it's not too much harder to reorganize headers in the future.

Notes for each file:
* **idle.c**: I changed gMainThread to sMainThread because it seems it should not be declared in a header, and adjusted block numbers.
* **z_actor.c**: It could maybe work to move `sCurCeilingPoly` and `sCurCeilingBgId` later in the file, but this caused BSS ordering problems for debug. Instead, I added some block number padding in the middle.
* **z_camera.c**: I had to add a large amount of padding between two in-function statics so I'm really unsure of how this was written originally. Padding in other places would also work.
* **z_collision_check.c**: Without the `sBssDummy` variables added for debug, the block numbers actually line up very nicely. I removed some scopes and moved a struct definition to make room.
* **z_common_data.c**: Added padding
* **z_demo.c**: Here it does seem like `D_8015FCCC`, `D_8015FCD0`, and `D_8015FCE4` should be global, and that the variables are declared in a different order than they're defined (maybe other ways are possible but it seems like it would require lots of padding)
* **z_kankyo.c**: This is definitely the jankiest file here. First, it seems like `gLensFlareUnused` should not be declared in a header, so I renamed it to `sLensFlareUnused`. Second, it seems like `gLightningStrike` should be declared before the other variables, but with some space in between before the other z_kankyo variables. I couldn't find a clean way to do this so I moved `sLightningFlashAlpha`, `sSunDepthTestX`, and `sSunDepthTestY` to the header too but left a TODO for later reorganization. Finally, I added a large amount of padding before `sNGameOverLightNode`. It could be moved farther down in the file, but this causes BSS ordering issues for debug.
* **z_kaleido_scope_call.c**: It seems like `D_8016139C` should be a global variable here.
* **z_play.c**: It seems like `sPlayVisMono` must be global here, and it would be easier if `sTransitionTile` was too. I think it makes sense given the similarity to `gVisMonoColor` and `gTransitionTileState` respectively
* **main.c**: Seems to require declaring variables in a different order in the header, and I moved `gPadMgr` to be closer to the others instead of in padmgr.h.
* **speed_meter.c**: It seems like `D_8016A578` should be a global variable here.
* **sys_math3d.c**: Very fragile, requires a specific starting block number.
* **fault.c**: The current placement of `gFaultMgr` makes this rather tight (it has block number 4 and the other variables must go before it), so I had to add different padding for retail and debug. This should be more lenient if the headers change.
* **fault_drawer.c**: Required no changes and should not be fragile (has 1/256 chance of getting the wrong ordering)

After #1917, #1918, #1921, and #1926, gc-eu-mq is now completely matched up to the end of code (other than the ROM checksum).

Dragorn421 reviewed (commented):
The "x should be block no y" comments are very nice

Not a fan of the s->g renames and moving declarations to variables.h, it's going backwards documentation-wise and eventually for cleanup the reverse should be done. If it's really the only way though...
The s prefix is for symbols that aren't global, and are used in the file they're declared in only
It is unrelated to the static qualifier though there is some conceptual similarity

cadmic commented:
For some files it's the only way, for others it's not. In a little bit I'll write more about which files need what

cadmic commented:
I reverted many of the s->g changes and added notes for each file in the PR description. I initially tried to put as many things in a header as possible because I thought that would be more robust to header changes, but I guess cleaner documentation is more important.

cadmic commented:
Since this hasn't gotten much feedback yet I pushed BSS ordering fixes for overlays too. Actually the main difficulty there is that we don't have many options for controlling the ordering without `-g3` which forces us to unnaturally declare variables at the top instead of spreading them out in the file. Spreading them out would make it easier to match the ordering in retail without tons of padding.

cadmic commented:
I merged main and this PR now builds OK

cadmic commented:
From discord: fig thinks the `INCREMENT_BLOCK_NUMBER_BY` macros are too noisy, maybe a `#pragma` and some kind of source preprocessor would be better

cadmic commented:
I introduced `tools/preprocess.py` which will progress `#pragma increment_block_number N` for all files in src. I made it do the reencoding too because it's already juggling temporary files and changing the encoding is very easy in Python.

This does slow the build down somewhat (by 10-20s with `-j8`) but maybe it's okay because that's basically how it was with asm-processor

Dragorn421 commented on tools/preprocess.py:
```diff
@@ -0,0 +1,63 @@
+
+# SPDX-FileCopyrightText: © 2024 ZeldaRET
+# SPDX-License-Identifier: CC0-1.0
+
+# Preprocess a C file to:
+# * Re-encode from UTF-8 to EUC-JP
+# * Replace `#pragma increment_block_number i` with `i` fake structs for controlling BSS ordering
+
+from pathlib import Path
```
Can you put a "usage" line similar to reencode.sh 's somewhere

Dragorn421 commented on tools/preprocess.py:
```diff
@@ -0,0 +1,63 @@
+        tmp.flush()
+
+        compile_command = sys.argv[1:-1] + ["-I", filename.parent, tmp.name]
+        process = subprocess.run(compile_command)
```
Running the command on the file while it's open for writing sounds sketchy, even with the flush

asm-processor's build.py handles this by creating a temporary folder, and then the file inside that, closing the file before running the compile command. And then deleting the temporary folder

We historically have had various issues with temp files in asm-processor (apparently the current version has no problems), so I'm a bit cautious

cadmic replied:
Changed to a tmpdir (which cleaned up ido_block_numbers.py a bit actually since the generated filename is no longer random)

fig02 commented:
Thanks for being flexible and trying out a new approach, appreciate it. I like the way this looks alot more

Dragorn421 reviewed (approved):
Don't worry about the slowdown :)

Dragorn421 reviewed (changes requested):
Sorry didn't mean to approve o_O

fig02 commented on src/code/z_collision_check.c:
```diff
@@ -2,11 +2,21 @@
 #include "terminal.h"
 #include "overlays/effects/ovl_Effect_Ss_HitMark/z_eff_ss_hitmark.h"
 
+typedef enum {
+    /* 0 */ MASSTYPE_IMMOVABLE,
+    /* 1 */ MASSTYPE_HEAVY,
+    /* 2 */ MASSTYPE_NORMAL
+} ColChkMassType;
```
I do definitley prefer having this near its localized usage. Is this the only way to achieve the desired effect, or could you also use increment_block_number to achieve it?

cadmic replied:
The block numbers work out perfectly without it which makes me think the enum definition was somewhere else in the original code (in a header maybe?). This adds 3 block numbers so we could move it back together with an `increment_block_number 253` to make it wrap all the way around, but I thought it was cleaner without that

fig02 replied:
This feels, to me, to be the same caliber as trying to match something like line numbers in the code. Feel like we should prioritize keeping things organizationally in the file where we like them. but more opinions needed probably

cadmic replied:
I don't feel strongly about it so I moved it back

fig02 commented on src/code/z_kankyo.c:
```diff
@@ -183,7 +188,7 @@ f32 sSandstormLerpScale = 0.0f;
 u8 gCustomLensFlareOn;
 Vec3f gCustomLensFlarePos;
-s16 gLensFlareUnused;
+s16 sLensFlareUnused;
 s16 gLensFlareScale;
 f32 gLensFlareColorIntensity;
 s16 gLensFlareGlareStrength;
```
this change doesnt make too much sense to me, cause the other lens flare variables are meant to be accessed globally to change its properties. I would assume that this variable followed that same pattern, when/if it actually meant something. But since its unused I guess I dont care too much

cadmic replied:
I think the BSS ordering suggests that this variable was not declared in a header (or at least, not in the same place as the other variables surrounding it here) since only `sLensFlareUnused` is reordered in retail

fig02 commented on src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -353,11 +353,24 @@ void Player_Action_80850E84(Player* this, PlayState* play);
+// For retail BSS ordering, the block number of sDogSpawnPos in Player_Update
+// must be between 0 and 53 inclusive.
+
+// TODO: There's probably a way to do this with less padding by spreading the variables out and moving
+// data around. It would be easier if we had more options for controlling BSS ordering in debug.
+#pragma increment_block_number 30
```
for the record I am very much in favor of spreading data around near its localized usages, and have been slowly doing this in player when I can, as I document. 
I can make a more concentrated effort to do this all at once if it would be helpful

cadmic replied:
The big problem with spreading the data around right now is that the block numbers don't work out for debug and we don't have a`#pragma increment_block_number` we can use there. Maybe it would work out better if the headers were split up more like they were in the original code?

fig02 replied:
do you mean when we untangle our global headers in general? or something more specific to player?

cadmic replied:
If the things that player needs are spread out in more headers, then more block numbers will be used (since each headers get assigned a block number if something is used from it) which might shift things enough in player so that the block numbers wrap around in the right place for spreading the data out to work. I don't remember the details for player though, like how many additional block numbers we need for debug

fig02 replied:
okay I understand. Yeah we'll reevaluate as we do more header untangling

cadmic commented on src/code/z_actor.c:
```diff
@@ -1894,7 +1897,9 @@ s32 func_8002F9EC(PlayState* play, Actor* actor, CollisionPoly* poly, s32 bgId,
+
+// Local data used for Farore's Wind light (stored in BSS)
 LightInfo D_8015BC00;
 LightNode* D_8015BC10;
```
I removed the comment because retail BSS ordering shows these are actually separate variables, there's other variables in between `D_8015BC10` and `D_8015BC14`, and `D_8015BC14` and `D_8015BC18` now (they're "interleaved" with `sCurCeilingPoly` and `sCurCeilingBgId`)

fig02 reviewed (approved):
more opinions could be given on https://github.com/zeldaret/oot/pull/1927#discussion_r1559868512 if anyone has thoughts.

But as-is this is fine for me
